nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-restapi-processors/src/test/java/org/apache/nifi/processors/elasticsearch/PutElasticsearchRecordTest.java

```diff
@@ -0,0 +1,782 @@
+            long trueScriptedUpsertCount = items.stream().filter(IndexOperationRequest::isScriptedUpsert).count();
+            long s = items.stream().filter(item ->  script.equals(item.getScript())).count();
+            long emptyDynamicTemplates = items.stream().filter(item ->  item.getDynamicTemplates().isEmpty()).count();
+            long dt = items.stream().filter(item ->  dynamicTemplates.equals(item.getDynamicTemplates())).count();
+            items.forEach(item -> {
+                assertNotNull(item.getId());
+                assertTrue(item.getId().startsWith("rec-"));
+                assertEquals("message", item.getType()); });
+
+            assertEquals(3, a);
+            assertEquals(3, b);
+            assertEquals(5, index);
+            assertEquals(1, create);
+            assertEquals(4, msg);
+            assertEquals(1, empties);
+            assertEquals(1, nulls);
+            assertEquals(1, timestamp);
+            assertEquals(5, timestampDefault);
+            assertEquals(0, ts);
+            assertEquals(0, id);
+            assertEquals(5, emptyScript);
+            assertEquals(5, falseScriptedUpsertCount);
+            assertEquals(1, trueScriptedUpsertCount);
+            assertEquals(1, s);
+            assertEquals(5, emptyDynamicTemplates);
+            assertEquals(1, dt);
+        });
+
+        registry.addSchema("recordPathTest", AvroTypeUtil.createSchema(new Schema.Parser().parse(newSchema)));
+
+        runner.removeProperty(PutElasticsearchRecord.INDEX_OP);
+        runner.setProperty(PutElasticsearchRecord.INDEX_OP_RECORD_PATH, "/op");
+        runner.setProperty(PutElasticsearchRecord.ID_RECORD_PATH, "/id");
+        runner.setProperty(PutElasticsearchRecord.INDEX_RECORD_PATH, "/index");
+        runner.setProperty(PutElasticsearchRecord.TYPE_RECORD_PATH, "/type");
+        runner.setProperty(PutElasticsearchRecord.AT_TIMESTAMP_RECORD_PATH, "/ts");
+        runner.setProperty(PutElasticsearchRecord.SCRIPT_RECORD_PATH, "/script");
+        runner.setProperty(PutElasticsearchRecord.SCRIPTED_UPSERT_RECORD_PATH, "/scripted_upsert");
+        runner.setProperty(PutElasticsearchRecord.DYNAMIC_TEMPLATES_RECORD_PATH, "/dynamic_templates");
+        runner.enqueue(Paths.get(TEST_DIR, "1_flowFileContents.json"), Collections.singletonMap("schema.name", "recordPathTest"));
+
+        runner.run();
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESS, 1);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILURE, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_RETRY, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_FAILED_RECORDS, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_SUCCESSFUL_RECORDS, 0);
+        runner.assertTransferCount(PutElasticsearchRecord.REL_ERROR_RESPONSES, 0);
+
+        runner.clearTransferState();
```

Comment:
It's probably worth taking this opportunity to split up these tests rather than keeping the really large `Test` scenarios that run the processor multiple times, with transfer state clearning in between - it would be easit to understand where failures are happening if we had everything in separate tests

Hopefully most of these can use a common `private` method that does the common setup/assertions, and simply set the test-specific properties/assertion logic

Reply:
N.B. there are several test methods in this class (and probably others) that should be considered for such refactoring - suggest searching for any instance of `clearTransferState` and considering whether the test(s) should be refactored

Reply:
@ChrisSamo632 So in test `testRecordPathFeatures` there are  8 `clearTransferState` calls. I understand you would like these split up into 9 separate unit tests. Any suggestions as to what these other tests should be called? It is not apparent to me what each of these scenarios are testing.

Reply:
I've added some suggestions to break up this Test Case, but also for others in this class (i.e. `testDateTimeFormatting` and `testErrorRelationship`)